src/game/MotionGenerators/PathFinder.cpp

```diff
@@ -175,7 +186,7 @@ void PathFinder::setArea(uint32 mapId, float x, float y, float z, uint32 area, f
 
     MMAP::MMapManager* mmap = MMAP::MMapFactory::createOrGetMMapManager();
 
-    dtNavMeshQuery const* query = mmap->GetNavMeshQuery(mapId, 0);
```

Comment:
I have not watched this code when it was initially merged since my codepath doesnt use setArea. If I am correct however, this is a massive race condition.

Reply:
Even in the new version

Reply:
> Even in the new version

Could you please elaborate? My assumption was that these lines are not much different from: https://github.com/cmangos/mangos-tbc/blob/1605ff06b177df6cdb4b4d72b7038ea273ca70d5/src/game/MotionGenerators/PathFinder.cpp#L46 which is called for bot-movement in similar fashion.

Do you have some advice how you'd prefer this functionality be implemented?

Reply:
That one is fine. But imagine this, a map, has multiple instances. Lets say BG. You have 3 wsgs happening side by side, with different instance ids, BUT the bots inside the bgs all access default instance 0. Its safe to access a mesh from different threads, its not safe to write to a mesh from 3 different threads (we can assume each map instance is own thread)

Reply:
Ok, if its only used during generation stage thats fine. Where is it first generated however? I assume whole map is generated at once right?

Reply:
> Ok, if its only used during generation stage thats fine. Where/when is it first generated however? I assume whole map is generated at once right?

Correct. The initial generation is done here:
https://github.com/cmangos/playerbots/blob/8aa5c3a903a41c6eec91bd0720c234ef82bf6a45/playerbot/TravelMgr.cpp#L942
This creates 1 thread for each mapId and stops the server until all calculations are done.


This is 'only' called when the server's ai_playerbot_travelnode table is empty. However this table is never empty for normal use since we distribute the content: https://github.com/cmangos/playerbots/blob/8aa5c3a903a41c6eec91bd0720c234ef82bf6a45/sql/world/tbc/ai_playerbot_travel_nodes.sql

So the generation stage is something only a developer will run to refresh/update the information of that .sql file.

Reply:
Okay then, disregard my concerns.